Added support to forcefully disable recording and/or alerting rules evaluation

pracucci opened the pull request:
#### What this PR does
I'm working on a tenants migration between two Mimir cluster and, as part of this migration, I need to selectively disable alerting rules for specific tenants. Instead of doing an hack in the ruler, I thought it could be generally useful having some advanced settings to forcefully disable recording and/or alerting rules evaluation on a per-tenant basis.

Note to reviewers:
- I think the only risky change in this PR is that the new function `filterRuleGroupsByEnabled()` manipulates the rules configuration in place. I've done it in place to avoid memory allocations at all. I checked both the bucket and local storage implementations and I don't think it will introduce issues, because both of them return a fresh map/slice each time. However, GEM also uses a custom rules manager and, even if I think it's not an issue today, this in-place manipulation could potentially introduce subtle bugs in the future. If you don't have concerns about introducing some memory allocations when rules evaluation is disabled, then I would suggest make some copies just to be safe.
- I did some refactoring in `TestRuler` to avoid using global variables and clean it up a bit

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented:
_Looking at the races reported by CI._

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+				&rulespb.RuleGroupDesc{
+					Name:      "group1",
+					Namespace: "namespace1",
+					User:      "user1",
```
```suggestion
					User:      userID,
```

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+				&rulespb.RuleGroupDesc{
+					Name:      "group1",
+					Namespace: "namespace1",
+					User:      "user1",
```
```suggestion
					User:      userID,
```

dimitarvdimitrov replied:
aside: what will happen if I forget to add a user to my test case? Should we verify that  `RuleGroupDesc.User == userID`  for each test case? Not sure how much of a safety net we need for contributors.

pracucci replied:
Currently there's no significant impact (it's used in few logs, and in part of the code not triggered by these tests). However, who know how code will change in the future? :) I will add a pre-condition check as you suggested.

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+				},
+			},
+			overrides: map[string]rulerLimits{
+				"user1": {
```
```suggestion
				userID: {
```

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+				&rulespb.RuleGroupDesc{
+					Name:      "group1",
+					Namespace: "namespace1",
+					User:      "user1",
```
```suggestion
					User:      userID,
```

dimitarvdimitrov replied:
i think there a few other places with this, can you please inplace the constant there too

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+					Namespace:     "namespace1",
+					User:          "user1",
+					SourceTenants: []string{"tenant-1"},
+					Rules: []*rulespb.RuleDesc{
+						{
+							Record: "UP_RULE",
+							Expr:   "up",
+						},
+						{
+							Alert: "UP_ALERT",
+							Expr:  "up < 1",
+						},
+					},
+					Interval: interval,
+				},
```
```suggestion
					Rules:     []*rulespb.RuleDesc{mockRecordingRuleDesc("UP_RULE", "up"), mockAlertingRuleDesc("UP_ALERT", "up < 1")},
					Interval:  interval,
```

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
+					Namespace:     "namespace1",
+					User:          "user1",
+					SourceTenants: []string{"tenant-1"},
+					Rules: []*rulespb.RuleDesc{
+						{
+							Record: "UP_RULE",
+							Expr:   "up",
+						},
+						{
+							Alert: "UP_ALERT",
+							Expr:  "up < 1",
+						},
+					},
+					Interval: interval,
```
```suggestion
					Rules:     []*rulespb.RuleDesc{mockRecordingRuleDesc("UP_RULE", "up"), mockAlertingRuleDesc("UP_ALERT", "up < 1")},
					Interval:  interval,
```

dimitarvdimitrov commented on pkg/ruler/api_test.go:
```diff
@@ -27,125 +27,219 @@ import (
-									Name:   "UP_ALERT",
-									Query:  "up < 1",
-									State:  "inactive",
-									Health: "unknown",
-									Type:   "alerting",
-									Alerts: []*Alert{},
-								},
-							},
-							Interval: 60,
+							Alert: "UP_ALERT",
+							Expr:  "up < 1",
+						},
+					},
```
```suggestion
					Rules:     []*rulespb.RuleDesc{mockRecordingRuleDesc("UP_RULE", "up"), mockAlertingRuleDesc("UP_ALERT", "up < 1")},
					Interval:  interval,
```

not sure if you consider this out of scope

dimitarvdimitrov commented on development/mimir-microservices-mode/config/mimir.yaml:
```diff
@@ -164,6 +164,7 @@ limits:
   max_global_exemplars_per_user: 5000
   query_sharding_total_shards: 16
   query_sharding_max_sharded_queries: 32
+  ingestion_rate: 50000
```
Did this slip through?

56quarters commented on pkg/ruler/ruler.go:
```diff
@@ -602,6 +606,93 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring r
+				}
+			}
+
+			// If all rules were removed then we should the whole group,
```
```suggestion
			// If all rules were removed then we should remove the whole group,
```

56quarters commented on pkg/ruler/ruler.go:
```diff
@@ -602,6 +606,93 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring r
+		}
+
+		if removedRulesTotal > 0 {
+			// We don't expect rules evaluation to be disabled for the normal use case. For this reason,
```
If we expect this to be rare does it make sense to return a new value in place instead of modifying the existing one to make the code a bit more straight forward? I know you mention avoiding any new allocations but it seems like we should prioritize clarity if this is indeed rare and would only be used during migrations.

pracucci replied:
Yeah, I share your concern too and I can't take a decision myself :) It's not common to use it, but if you use it in a Mimir cluster with a large number of tenants / rules, then there could be a bit of impact if we reallocate all data structures instead of doing a in-place editing.

pracucci replied:
I went for the copy version. See: https://github.com/grafana/mimir/pull/3088#issuecomment-1263352748

dimitarvdimitrov reviewed (approved):
LGTM, with a few comments around the tests.


I don't think the inplace modification of slices is a problem. `loadRuleGroups` also modifies rule groups in place, so I don't see how this makes the situation different (maybe because it resizes slices?)

56quarters reviewed (approved):
LGTM, just one non-blocking comment.

pracucci commented on pkg/ruler/ruler_test.go:
```diff
@@ -76,30 +76,75 @@ func defaultRulerConfig(t testing.TB) Config {
 
-func (r ruleLimits) RulerTenantShardSize(_ string) int {
-	return r.tenantShard
+func newMockLimits() *mockLimits {
+	return &mockLimits{
+		defaults: rulerLimits{
+			maxRuleGroups:                   20,
+			maxRulesPerRuleGroup:            15,
+			alertingRulesEvaluationEnabled:  true,
+			recordingRulesEvaluationEnabled: true,
+		},
+		overrides: map[string]rulerLimits{},
+	}
```
Let's do a step back and try to address in a prequel PR: https://github.com/grafana/mimir/pull/3090

pracucci commented:
@dimitarvdimitrov I should have addressed all your comments, and I rebased `main` after merging https://github.com/grafana/mimir/pull/3090. Could you take another look, please?

pracucci commented:
I went for the copy version of `filterRuleGroupsByEnabled()`. It's safer, easier to reason about, and the benchmark compared to the in-place version doesn't show a big difference. Reason is that all slices elements are pointers, and we never change them (just filter out), so all in all we have just to re-allocate slices of pointers in the copy version.

```
name                                                   old time/op    new time/op    delta
FilterRuleGroupsByEnabled/all_rules_disabled-12          45.2ms ±32%    33.2ms ± 2%    ~     (p=0.239 n=3+3)
FilterRuleGroupsByEnabled/all_rules_enabled-12           33.2ms ± 2%    32.3ms ± 1%    ~     (p=0.125 n=3+3)
FilterRuleGroupsByEnabled/recording_rules_disabled-12    37.7ms ± 1%    39.4ms ± 2%  +4.71%  (p=0.016 n=3+3)
FilterRuleGroupsByEnabled/alerting_rules_disabled-12     41.5ms ± 6%    39.1ms ± 1%    ~     (p=0.197 n=3+3)

name                                                   old alloc/op   new alloc/op   delta
FilterRuleGroupsByEnabled/all_rules_disabled-12          27.2MB ± 0%    27.3MB ± 0%  +0.36%  (p=0.000 n=3+3)
FilterRuleGroupsByEnabled/all_rules_enabled-12           27.2MB ± 0%    27.2MB ± 0%    ~     (p=0.751 n=3+3)
FilterRuleGroupsByEnabled/recording_rules_disabled-12    27.6MB ± 0%    29.9MB ± 0%  +8.17%  (p=0.000 n=3+3)
FilterRuleGroupsByEnabled/alerting_rules_disabled-12     27.6MB ± 0%    29.9MB ± 0%  +8.17%  (p=0.000 n=3+3)

name                                                   old allocs/op  new allocs/op  delta
FilterRuleGroupsByEnabled/all_rules_disabled-12            433k ± 0%      433k ± 0%  +0.00%  (p=0.025 n=3+3)
FilterRuleGroupsByEnabled/all_rules_enabled-12             433k ± 0%      433k ± 0%    ~     (p=1.000 n=3+3)
FilterRuleGroupsByEnabled/recording_rules_disabled-12      438k ± 0%      459k ± 0%  +4.80%  (p=0.000 n=3+3)
FilterRuleGroupsByEnabled/alerting_rules_disabled-12       438k ± 0%      459k ± 0%  +4.80%  (p=0.000 n=3+3)
```

pracucci commented:
I've redone all manual tests and looks good.

dimitarvdimitrov reviewed (approved):
LGTM thanks for addressing my feedback

56quarters reviewed (approved):
Thanks for making the copy/modify changes!
